### PR TITLE
Escape before print MetadataMap which could contain invalid utf8 strings.

### PR DESCRIPTION
**Pull request by stevenzzzz**
Commit Message: Escape before print MetadataMap which could contain invalid utf8 strings.
Additional Description:
Risk Level: LOW 
Testing: unit test
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #22609
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/common/http/metadata_interface_test.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using ::testing::HasSubstr;
+
+namespace Envoy {
+namespace Http {
+namespace {
+
+TEST(MetadataMapTest, KeyValueEscaped) {
```
Comment (adisuissa) [inline test/common/http/metadata_interface_test.cc:19]: IIUC this test only validates the usage of "operator<<".
Can you add a test to validate that the formatter works properly? (maybe something that uses [`LogLevelSetter`](https://github.com/envoyproxy/envoy/blob/972b33265f26ed206c047fd05d9427f4cc6408aa/test/test_common/logging.h#L31) and [`EXPECT_LOG_CONTAINS`](https://github.com/envoyproxy/envoy/blob/main/test/test_common/logging.h#L163) will do the trick).

**Review by adisuissa — CHANGES_REQUESTED** [review]
Thanks for fixing this!

I think that escaping all strings should be better in the long-run. Adding the following formatter (to `source/common/common/fmt.h`) should work:
```
template <> struct formatter<std::string> : formatter<string_view> {
  auto format(const std::string& str_in, fmt::format_context& ctx) -> decltype(ctx.out()) {
    string_view fmt_string_view(absl::CEscape(str_in));
    return formatter<string_view>::format(fmt_string_view, ctx);
  }
};
```

/wait

**Comment by stevenzzzz** [comment]
> auto format(const std::string& str_in, fmt::format_context& ctx) -> decltype(ctx.out()) {

just this change it breaks the unit test I created. 

I am not sure we should escaping all the strings, but maybe we should as this is used in just logging.
@alyssawilk

File: test/common/http/metadata_interface_test.cc (hunk 1 above)
Comment (stevenzzzz) [reply test/common/http/metadata_interface_test.cc:19]: it also validates that the ENVOY_LOG_MISC is not crashing, which, without this PR would crash.

**Comment by adisuissa** [comment]
cc @yanavlasov

**Comment by dyfrgi** [comment]
Rather than create a new specialized formatter, we could use fmt's built-in ostream formatter to re-use the `operator<<()` we already have for `MetadataMap`. e.g.
```
template <> struct fmt::formatter<MetadataMap> : fmt::ostream_formatter {};
```
Idea from https://fmt.dev/latest/api.html#std-ostream-support. It's defined in fmt/ostream.h, so we'd need to include that in metadata_interface.h.

**Comment by adisuissa** [comment]
> Rather than create a new specialized formatter, we could use fmt's built-in ostream formatter to re-use the `operator<<()` we already have for `MetadataMap`. e.g.
> 
> ```
> template <> struct fmt::formatter<MetadataMap> : fmt::ostream_formatter {};
> ```
> 
> Idea from https://fmt.dev/latest/api.html#std-ostream-support. It's defined in fmt/ostream.h, so we'd need to include that in metadata_interface.h.

Yes. The API you mentioned is from FMT-LIB v9, and we use the same default formatter (that uses ostream-formatter which is essentially the default_formatter) that calls operator<<. The include line [here](https://github.com/envoyproxy/envoy/blob/main/envoy/registry/registry.h#L22) creates a new formatter for the MetadataMap and overrides the ostream one.

**Comment by stevenzzzz** [comment]
I dont think we can tell when the  fmt/ranges.h header will be included (either by Envoy or any lib Envoy depends on), which result in the map-specialization been expanded/used. In the write_range_entry call, the invalid utf8 string leads to ASAn error.  I think we can prevent it from happening before that's fixed in fmtlib.

File: envoy/http/metadata_interface.h [hunk 2]
```diff
@@ -2,10 +2,13 @@
 
 #include <functional>
 #include <iostream>
+#include <sstream>
 #include <memory>
 #include <vector>
 
 #include "absl/container/node_hash_map.h"
+#include "absl/strings/escaping.h"
+#include "fmt/ranges.h"
```
Comment (adisuissa) [inline envoy/http/metadata_interface.h:11]: nit: should this be changed to "fmt/format.h" or "fmt/stream.h"?

File: test/common/http/metadata_interface_test.cc [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#include <array>
+#include <cstdint>
+#include <string>
+
+#include "envoy/http/metadata_interface.h"
+
+#include "source/common/common/fmt.h"
+#include "source/common/common/logger.h"
+#include "source/common/http/utility.h"
```
Comment (adisuissa) [inline test/common/http/metadata_interface_test.cc:9]: This should be: "fmt/ranges.h", and please add a comment why it is added.

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a couple small nits.

@mattklein123 PTAL for a non-googler review

File: envoy/http/metadata_interface.h (hunk 2 above)
Comment (stevenzzzz) [reply envoy/http/metadata_interface.h]: change to format.h.

File: test/common/http/metadata_interface_test.cc (hunk 3 above)
Comment (stevenzzzz) [reply test/common/http/metadata_interface_test.cc]: done

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**